planner/core/optimizer.go

```diff
@@ -148,6 +148,36 @@ func postOptimize(sctx sessionctx.Context, plan PhysicalPlan) PhysicalPlan {
 	plan = eliminatePhysicalProjection(plan)
 	plan = injectExtraProjection(plan)
 	plan = eliminateUnionScanAndLock(sctx, plan)
+	plan = enableParallelApply(sctx, plan)
+	return plan
+}
+
+func enableParallelApply(sctx sessionctx.Context, plan PhysicalPlan) PhysicalPlan {
+	if !sctx.GetSessionVars().EnableParallelApply {
+		return plan
+	}
+	// the parallel apply has three limitation:
+	// 1. the parallel implementation now cannot keep order;
+	// 2. the inner child has to support clone;
+	// 3. if one Apply is in inner-side of another Apply, it cannot be parallel, for example:
+	//		The topology of 3 Apply operators are A1(A2, A3), which means A2 is the outer child of A1
+	//		while A3 is the inner child. Then A1 and A2 can be parallel and A3 cannot.
+	if apply, ok := plan.(*PhysicalApply); ok {
+		outerIdx := 1 - apply.InnerChildIdx
+		noOrder := len(apply.GetChildReqProps(outerIdx).Items) == 0 // limitation 1
+		_, err := SafeClone(apply.Children()[apply.InnerChildIdx])
+		supportClone := err == nil // limitation 2
+		if noOrder && supportClone {
+			apply.Concurrency = sctx.GetSessionVars().ExecutorConcurrency
+		}
+
+		apply.SetChild(outerIdx, enableParallelApply(sctx, apply.Children()[outerIdx]))
+		// apply.SetChild(innerIdx, enableParallelApply(innerChild)) <--- limitation 3
```

Comment:
why annotate it？

Reply:
It means that because of limitation 3, we cannot invoke `enableParallelApply` recursively for the inner child.
I have removed this annotated line and added more comments about it. PTAL.